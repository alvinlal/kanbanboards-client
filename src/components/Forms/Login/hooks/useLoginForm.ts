import { yupResolver } from '@hookform/resolvers/yup';
import { AxiosError } from 'axios';
import { useState, useCallback } from 'react';
import { useForm, SubmitHandler, Control, FieldErrorsImpl } from 'react-hook-form';
import { useNavigate } from 'react-router-dom';
import { useAuth } from '../../../../hooks/useAuth/useAuth';
import { useErrorHandlers } from '../../../../hooks/useErrorHandlers/useErrorHandlers';
import { UnauthorizedException } from '../../../../types/UnauthorizedException';

import { loginSchema } from '../validators/loginSchema';

interface UseLoginForm {
  loading: boolean;
  isValid: boolean;
  errors: Partial<FieldErrorsImpl<LoginFormInputs>>;
  control: Control<LoginFormInputs, unknown>;
  handleContinueWithGoogle: () => void;
  submitForm: (e: React.FormEvent) => void;
}

export type LoginFormInputs = {
  email: string;
  password: string;
};

export const useLoginForm = (): UseLoginForm => {
  const {
    control,
    handleSubmit,
    setError,
    formState: { errors, isValid },
  } = useForm<LoginFormInputs>({
    mode: 'onChange',
    resolver: yupResolver(loginSchema),
  });
  const [loading, setLoading] = useState(false);
  const { logIn } = useAuth();
  const { handleRequestError } = useErrorHandlers();
  const navigate = useNavigate();

  const onSubmit: SubmitHandler<LoginFormInputs> = useCallback(
    async ({ email, password }) => {
      try {
        setLoading(true);
        await logIn({ email, password });
        navigate('/');
      } catch (err) {
        const requestErrors = handleRequestError(err as AxiosError) as UnauthorizedException;

        if (requestErrors) {
          setError('password', {
            message: requestErrors.message,
          });
        }
      } finally {
        setLoading(false);
      }
    },
    [logIn, navigate, handleRequestError, setError]
  );

  const handleContinueWithGoogle = useCallback(() => {
    window.location.href = `${import.meta.env.VITE_API_ENDPOINT}/auth/google`;
  }, []);

  const submitForm = (e: React.FormEvent) => {
    e.preventDefault();
    handleSubmit(onSubmit)();
  };

  return {
    control,
    errors,
    isValid,
    loading,
    submitForm,
    handleContinueWithGoogle,
  };
};
