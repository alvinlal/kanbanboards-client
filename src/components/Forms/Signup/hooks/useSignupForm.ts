import { yupResolver } from '@hookform/resolvers/yup';
import { AxiosError } from 'axios';
import { useState, useCallback } from 'react';
import {
  useForm,
  SubmitHandler,
  FieldErrorsImpl,
  Control,
  UseFormHandleSubmit,
  UseFormTrigger,
} from 'react-hook-form';
import { useNavigate } from 'react-router-dom';
import useAuth from '../../../../hooks/useAuth';
import useErrorHandlers from '../../../../hooks/useErrorHandlers';
import signupSchema from '../validators/signupSchema';

interface UseSignupForm {
  loading: boolean;
  isValid: boolean;
  errors: Partial<FieldErrorsImpl<FormInputs>>;
  control: Control<FormInputs, unknown>;
  handleGoogleSignUp: () => void;
  handleSubmit: UseFormHandleSubmit<FormInputs>;
  onSubmit: SubmitHandler<FormInputs>;
  trigger: UseFormTrigger<FormInputs>;
}

type FormInputs = {
  email: string;
  password: string;
  confirmPassword: string;
};

export const useSignupForm = (): UseSignupForm => {
  const {
    control,
    handleSubmit,
    trigger,
    setError,
    formState: { errors, isValid },
  } = useForm<FormInputs>({
    mode: 'onChange',
    resolver: yupResolver(signupSchema),
  });
  const [loading, setLoading] = useState(false);
  const { signUp } = useAuth();
  const { handleRequestError } = useErrorHandlers();
  const navigate = useNavigate();

  const onSubmit: SubmitHandler<FormInputs> = useCallback(
    async ({ email, password }) => {
      try {
        setLoading(true);
        await signUp({ email, password });
        navigate('/');
      } catch (err) {
        const requestErrors = handleRequestError(err as AxiosError);
        if (requestErrors) {
          setError(
            'email',
            {
              message: requestErrors.message[0].constraints.isEmailExists,
            },
            {
              shouldFocus: true,
            }
          );
        }
      } finally {
        setLoading(false);
      }
    },
    [signUp, navigate, handleRequestError, setError]
  );

  const handleGoogleSignUp = useCallback(() => {
    window.location.href = `${import.meta.env.VITE_API_ENDPOINT}/auth/google`;
  }, []);

  return {
    control,
    errors,
    handleSubmit,
    handleGoogleSignUp,
    isValid,
    loading,
    onSubmit,
    trigger,
  };
};
