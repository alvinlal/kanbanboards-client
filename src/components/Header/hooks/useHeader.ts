import { useAuth } from '../../../hooks/useAuth/useAuth';
import { useToggleLeftSideNav } from '../../../hooks/useToggleLeftSideNav/useToggleLeftSideNav';

interface UseHeader {
  toggleLeftSideNav: () => void;
  logOut: () => Promise<void>;
}

export const useHeader = (): UseHeader => {
  const { logOut } = useAuth();
  const { toggleLeftSideNav } = useToggleLeftSideNav();

  return {
    toggleLeftSideNav,
    logOut,
  };
};
