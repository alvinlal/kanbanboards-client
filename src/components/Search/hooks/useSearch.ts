import throttle from 'lodash.throttle';
import React, {
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState,
} from 'react';
import search from '../../../api/search';
import SearchResponse from '../../../api/search/types/SearchResponse';

interface UseSearch {
  results: SearchResponse | null;
  isResultsVisible: boolean;
  setIsResultsVisible: React.Dispatch<React.SetStateAction<boolean>>;
  loading: boolean;
  searchInputRef: React.RefObject<HTMLInputElement>;
  throttledHandleChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
}

export const useSearch = (): UseSearch => {
  const [results, setResults] = useState<SearchResponse | null>(null);
  const [isResultsVisible, setIsResultsVisible] = useState(false);
  const [loading, setLoading] = useState(false);
  const searchInputRef = useRef<HTMLInputElement>(null);
  const abortController = useMemo(() => new AbortController(), []);

  const handleChange = useCallback(
    async (e: React.ChangeEvent<HTMLInputElement>) => {
      if (!e.target.value) {
        setResults(null);
        return;
      }
      try {
        setLoading(true);
        const { data } = await search(e.target.value, abortController.signal);
        if (searchInputRef.current?.value) {
          setResults(data);
        }
      } catch (err) {
        // send error to monitoring service
      } finally {
        setLoading(false);
      }
    },
    [abortController.signal]
  );

  const throttledHandleChange = useMemo(
    () => (e: React.ChangeEvent<HTMLInputElement>) => {
      const throttledFn = throttle(handleChange, 300);
      throttledFn(e);
    },
    [handleChange]
  );

  useEffect(() => {
    return () => {
      abortController.abort();
    };
  }, [abortController]);

  return {
    results,
    isResultsVisible,
    setIsResultsVisible,
    loading,
    searchInputRef,
    throttledHandleChange,
  };
};
