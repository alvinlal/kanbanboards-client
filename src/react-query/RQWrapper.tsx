/* eslint-disable no-console */
// this wrapper is for testing only

import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import React from 'react';

const queryClient = new QueryClient({
  logger: {
    log: console.log,
    warn: console.warn,
    error: () => {},
  },
});
export const RQWrapper: React.FC<{ children: React.ReactNode }> = ({ children }) => (
  <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
);
