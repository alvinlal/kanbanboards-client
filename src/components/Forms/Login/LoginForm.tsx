import { Controller } from 'react-hook-form';
import Button from '../../Button/Button';
import TextField from '../../Inputs/TextField/TextField';
import Spinner from '../../Spinner/Spinner';
import { useLoginForm } from './hooks/useLoginForm';
import styles from './LoginForm.module.scss';
import { ReactComponent as GoogleLogo } from '../../../assets/icons/google.svg';

type LoginFormProps = React.FormHTMLAttributes<HTMLFormElement>;

const LoginForm: React.FC<LoginFormProps> = (props) => {
  const { errors, isValid, handleContinueWithGoogle, control, loading, submitForm } =
    useLoginForm();

  return (
    <form className={styles.login__form} onSubmit={submitForm} {...props}>
      <h1>Login</h1>
      <Controller
        control={control}
        name="email"
        render={({ field: { ref, onChange, onBlur } }) => (
          <TextField
            autoFocus
            aria-label="email"
            type="email"
            placeholder="email"
            error={!!errors.email}
            onChange={onChange}
            onBlur={onBlur}
            ref={ref}
          />
        )}
      />
      <div className={styles.error}>{errors.email && <p>{errors.email.message}</p>}</div>
      <Controller
        control={control}
        name="password"
        render={({ field: { ref, onChange, onBlur } }) => (
          <TextField
            type="password"
            aria-label="password"
            placeholder="password"
            error={!!errors.password}
            onChange={onChange}
            onBlur={onBlur}
            ref={ref}
          />
        )}
      />
      <div className={styles.error}>{errors.password && <p>{errors.password.message}</p>}</div>
      <Button
        type="submit"
        style={{ width: '100%', height: '56px', borderRadius: 5 }}
        variant="primary"
        disabled={!isValid || loading}
      >
        <p>LOGIN</p>
        {loading && <Spinner />}
      </Button>
      <hr />
      <Button
        type="button"
        onClick={handleContinueWithGoogle}
        style={{ width: '100%', height: '56px', borderRadius: 5 }}
        variant="secondary"
      >
        <p>CONTINUE WITH GOOGLE</p>
        <GoogleLogo width={28} height={28} />
      </Button>
    </form>
  );
};

export default LoginForm;
