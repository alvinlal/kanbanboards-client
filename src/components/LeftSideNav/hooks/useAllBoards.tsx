import { useQuery } from '@tanstack/react-query';
import { getAllBoards } from '../../../api/board';
import { AllBoardsResponseDto } from '../../../api/board/dto/response/AllBoardsResponse.dto';
import queryKeys from '../../../react-query/queryKeys';

interface UseAllBoards {
  allBoards: AllBoardsResponseDto | undefined;
  isLoading: boolean;
  isError: boolean;
}

export const useAllBoards = (): UseAllBoards => {
  const { data: allBoards, isLoading, isError } = useQuery([queryKeys.allBoards], getAllBoards);

  return {
    allBoards,
    isLoading,
    isError,
  };
};
