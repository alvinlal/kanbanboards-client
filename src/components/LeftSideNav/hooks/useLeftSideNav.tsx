import { useMemo } from 'react';
import { AllBoardsResponse } from '../../../api/board/dto/response/AllBoardsResponse.dto';
import { useAllBoards } from '../../../hooks/useAllBoards/useAllBoards';
import { useAuth } from '../../../hooks/useAuth/useAuth';
import { useToggleLeftSideNav } from '../../../hooks/useToggleLeftSideNav/useToggleLeftSideNav';
import SkeletonBlock from '../../SkeletonBlock/SkeletonBlock';

interface UseLeftSideNav {
  LoadingBlocks: React.ReactNode[];
  allBoards: AllBoardsResponse | undefined;
  isLoading: boolean;
  isError: boolean;
  isLeftSideNavOpen: boolean;
  toggleLeftSideNav: () => void;
  logOut: () => Promise<void>;
}

export const useLeftSideNav = (): UseLeftSideNav => {
  const { allBoards, isLoading, isError } = useAllBoards();
  const { isLeftSideNavOpen, toggleLeftSideNav } = useToggleLeftSideNav();
  const { logOut } = useAuth();

  const LoadingBlocks = useMemo(() => {
    const blocks = [];

    for (let i = 0; i < 6; i += 1) {
      blocks.push(<SkeletonBlock width="200px" height="30px" key={i} />);
    }

    return blocks;
  }, []);

  return {
    LoadingBlocks,
    allBoards,
    isLoading,
    isError,
    isLeftSideNavOpen,
    toggleLeftSideNav,
    logOut,
  };
};
