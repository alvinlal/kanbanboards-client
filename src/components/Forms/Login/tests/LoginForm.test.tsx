/* eslint-disable @typescript-eslint/ban-ts-comment */
/**
 * Should show  error on empty email input
 * Should show error on empty password input
 * Should show error on incorrect credentials
 * Login button should be disabled if there are errors
 * Should show loading spinner on clicking login
 * Should not show any errors if all the fields are valid
 * Should navigate user to home page after successfull login
 * Should match snapshot
 * Should match loading snapshot
 * Should match error snapshot
 */

import userEvent from '@testing-library/user-event';
import { render, screen, waitFor } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { MemoryRouter, Router } from 'react-router-dom';
import { faker } from '@faker-js/faker';
import { createMemoryHistory } from 'history';
import { rest, server } from '../../../../../test-utils/msw/server';
import { LoginFormInputs } from '../hooks/useLoginForm';
import LoginForm from '../LoginForm';
import { RQWrapper } from '../../../../react-query/RQWrapper';
import { apiEndPoint } from '../../../../../test-utils/msw/baseUrls';
import defaultHandlers from '../../../../../test-utils/msw/defaultHandlers';
import { loginController } from '../../../../../test-utils/msw/handlers/auth/controllers/loginController';

describe('LoginForm.tsx', () => {
  const typeIntoForm = async ({ email, password }: Partial<LoginFormInputs>) => {
    const user = userEvent.setup();
    const emailInputElement = screen.getByLabelText('email') as HTMLInputElement;

    const passwordInputElement = screen.getByLabelText('password') as HTMLInputElement;

    if (email) {
      await user.type(emailInputElement, email);
    }

    if (password) {
      await user.type(passwordInputElement, password);
    }

    return {
      emailInputElement,
      passwordInputElement,
    };
  };

  const clickOnLoginButton = async () => {
    const loginButtonElement = screen.getByRole('button', {
      name: /LOGIN/i,
    }) as HTMLButtonElement;
    await userEvent.click(loginButtonElement);
    return loginButtonElement;
  };

  const clickOnGoogleButton = async () => {
    const googleButtonElement = screen.getByRole('button', {
      name: /CONTINUE WITH GOOGLE/i,
    }) as HTMLButtonElement;
    await userEvent.click(googleButtonElement);
    return googleButtonElement;
  };

  describe('Should show email errors', () => {
    describe('when email is empty', () => {
      it('shows correct error message', async () => {
        render(
          <MemoryRouter>
            <LoginForm />
          </MemoryRouter>,
          {
            wrapper: RQWrapper,
          }
        );

        const { emailInputElement } = await typeIntoForm({
          email: faker.internet.email(),
        });
        await userEvent.clear(emailInputElement);
        const emailIsRequiredErrorElement = await screen.findByText(/please enter your email/i);
        expect(emailIsRequiredErrorElement).toBeInTheDocument();
      });
    });

    describe('when email or password is incorrect', () => {
      it('should show correct error message', async () => {
        server.use(rest.post(apiEndPoint('/auth/login'), loginController[401]));
        render(
          <MemoryRouter>
            <LoginForm />
          </MemoryRouter>,
          {
            wrapper: RQWrapper,
          }
        );

        await typeIntoForm({
          email: faker.internet.email(),
          password: faker.internet.password(6),
        });

        await clickOnLoginButton();
        const incorrectCredentialsErrorElement = await screen.findByText(
          'incorrect email or password'
        );
        expect(incorrectCredentialsErrorElement).toBeInTheDocument();
      });
    });
  });

  describe('Should show password errors', () => {
    describe('when password is empty', () => {
      it('shows correct error message', async () => {
        render(
          <MemoryRouter>
            <LoginForm />
          </MemoryRouter>,
          { wrapper: RQWrapper }
        );
        const { passwordInputElement } = await typeIntoForm({
          password: faker.internet.password(),
        });
        await userEvent.clear(passwordInputElement);
        const passwordIsRequiredErrorElement = await screen.findByText(
          /please enter your password/i
        );
        expect(passwordIsRequiredErrorElement).toBeInTheDocument();
      });
    });
  });
  describe('Should show correct button states for signup button', () => {
    describe('when there is no errors', () => {
      it('signup button should not be disabled', async () => {
        render(
          <MemoryRouter>
            <LoginForm />
          </MemoryRouter>,
          { wrapper: RQWrapper }
        );

        await typeIntoForm({ email: faker.internet.email(), password: faker.internet.password(6) });
        const loginButtonElement = await clickOnLoginButton();
        await waitFor(() => {
          expect(loginButtonElement.disabled).toBe(false);
        });
      });
    });

    describe('when there are errors', () => {
      it('signup button should be disabled', async () => {
        render(
          <MemoryRouter>
            <LoginForm />
          </MemoryRouter>,
          { wrapper: RQWrapper }
        );

        await typeIntoForm({ email: faker.internet.email() });
        const loginButtonElement = await clickOnLoginButton();
        await waitFor(() => {
          expect(loginButtonElement.disabled).toBe(true);
        });
      });
    });

    describe('Should navigate user to the correct destination after successfull login', () => {
      describe('when login button is clicked with correct form inputs', () => {
        it('should navigate user to home page', async () => {
          const history = createMemoryHistory();
          history.push = jest.fn();

          render(
            <Router location={history.location} navigator={history}>
              <LoginForm />
            </Router>,
            { wrapper: RQWrapper }
          );

          await typeIntoForm({
            email: faker.internet.email(),
            password: faker.internet.password(6),
          });

          await clickOnLoginButton();

          expect(history.push).toHaveBeenCalledWith(
            { pathname: '/', search: '', hash: '' },
            undefined,
            {}
          );
        });

        it('should show loading spinner in login button', async () => {
          server.use(rest.post(apiEndPoint('/auth/login'), defaultHandlers.LOADING));
          render(
            <MemoryRouter>
              <LoginForm />
            </MemoryRouter>,
            { wrapper: RQWrapper }
          );

          await typeIntoForm({
            email: faker.internet.email(),
            password: faker.internet.password(6),
          });
          await clickOnLoginButton();
          const spinner = await screen.findByTestId('spinner');
          expect(spinner).toBeInTheDocument();
        });

        it('login button should be disabled while signup api request is loading', async () => {
          server.use(rest.post(apiEndPoint('/auth/login'), defaultHandlers.LOADING));
          render(
            <MemoryRouter>
              <LoginForm />
            </MemoryRouter>,
            { wrapper: RQWrapper }
          );

          await typeIntoForm({
            email: faker.internet.email(),
            password: faker.internet.password(6),
          });
          const signupButtonElement = await clickOnLoginButton();
          await waitFor(() => {
            expect(signupButtonElement.disabled).toBe(true);
          });
        });
      });

      describe('when continue with google button is clicked', () => {
        it('should navigate user to correct url', async () => {
          const assignMock = jest.fn();

          // @ts-ignore
          delete window.location;
          // @ts-ignore
          window.location = { assign: assignMock };

          render(
            <MemoryRouter>
              <LoginForm />
            </MemoryRouter>,
            { wrapper: RQWrapper }
          );

          await clickOnGoogleButton();
          expect(window.location.href).toBe(`${import.meta.env.VITE_API_ENDPOINT}/auth/google`);
        });
      });
    });
  });

  describe('Should match snapshots', () => {
    it("match's default snapshot", () => {
      const tree = renderer.create(
        <MemoryRouter>
          <RQWrapper>
            <LoginForm />
          </RQWrapper>
        </MemoryRouter>
      );
      expect(tree).toMatchSnapshot();
    });

    it("match's loading state snapshot", async () => {
      server.use(rest.post(apiEndPoint('/auth/login'), defaultHandlers.LOADING));
      const { asFragment } = render(
        <MemoryRouter>
          <LoginForm />
        </MemoryRouter>,
        {
          wrapper: RQWrapper,
        }
      );

      await typeIntoForm({
        email: faker.internet.email(),
        password: faker.internet.password(),
      });
      await clickOnLoginButton();

      await waitFor(async () => {
        const spinner = await screen.findByTestId('spinner');
        expect(spinner).toBeInTheDocument();
      });
      expect(asFragment()).toMatchSnapshot();
    });

    it("match's field error state snapshot", async () => {
      const { asFragment } = render(
        <MemoryRouter>
          <LoginForm />
        </MemoryRouter>,
        {
          wrapper: RQWrapper,
        }
      );

      const { emailInputElement, passwordInputElement } = await typeIntoForm({
        email: faker.internet.email(),
        password: faker.internet.password(),
      });

      await userEvent.clear(emailInputElement);
      await userEvent.clear(passwordInputElement);

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
