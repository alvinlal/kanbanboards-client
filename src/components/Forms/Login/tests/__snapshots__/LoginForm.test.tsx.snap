// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm.tsx Should match snapshots match's default snapshot 1`] = `
<form
  className="login__form"
  onSubmit={[Function]}
>
  <h1>
    Login
  </h1>
  <input
    aria-label="email"
    autoFocus={true}
    className="textfield null"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="email"
    type="email"
  />
  <div
    className="error"
  />
  <input
    aria-label="password"
    className="textfield null"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="password"
    type="password"
  />
  <div
    className="error"
  />
  <button
    className="button primary "
    disabled={true}
    style={
      {
        "height": "56px",
        "width": "100%",
      }
    }
    type="submit"
  >
    <p>
      LOGIN
    </p>
  </button>
  <hr />
  <button
    className="button secondary "
    onClick={[Function]}
    style={
      {
        "height": "56px",
        "width": "100%",
      }
    }
    type="button"
  >
    <p>
      CONTINUE WITH GOOGLE
    </p>
    <div
      height={28}
      width={28}
    />
  </button>
</form>
`;

exports[`LoginForm.tsx Should match snapshots match's field error state snapshot 1`] = `
<DocumentFragment>
  <form
    class="login__form"
  >
    <h1>
      Login
    </h1>
    <input
      aria-label="email"
      class="textfield error"
      placeholder="email"
      type="email"
    />
    <div
      class="error"
    >
      <p>
        please enter your email
      </p>
    </div>
    <input
      aria-label="password"
      class="textfield error"
      placeholder="password"
      type="password"
    />
    <div
      class="error"
    >
      <p>
        please enter your password
      </p>
    </div>
    <button
      class="button primary "
      disabled=""
      style="width: 100%; height: 56px;"
      type="submit"
    >
      <p>
        LOGIN
      </p>
    </button>
    <hr />
    <button
      class="button secondary "
      style="width: 100%; height: 56px;"
      type="button"
    >
      <p>
        CONTINUE WITH GOOGLE
      </p>
      <div
        height="28"
        width="28"
      />
    </button>
  </form>
</DocumentFragment>
`;

exports[`LoginForm.tsx Should match snapshots match's loading state snapshot 1`] = `
<DocumentFragment>
  <form
    class="login__form"
  >
    <h1>
      Login
    </h1>
    <input
      aria-label="email"
      class="textfield null"
      placeholder="email"
      type="email"
    />
    <div
      class="error"
    />
    <input
      aria-label="password"
      class="textfield null"
      placeholder="password"
      type="password"
    />
    <div
      class="error"
    />
    <button
      class="button primary "
      disabled=""
      style="width: 100%; height: 56px;"
      type="submit"
    >
      <p>
        LOGIN
      </p>
      <div
        class="spinner"
        data-testid="spinner"
      />
    </button>
    <hr />
    <button
      class="button secondary "
      style="width: 100%; height: 56px;"
      type="button"
    >
      <p>
        CONTINUE WITH GOOGLE
      </p>
      <div
        height="28"
        width="28"
      />
    </button>
  </form>
</DocumentFragment>
`;
